src/runtime_src/core/pcie/windows/device_windows.cpp

```diff
@@ -835,6 +835,83 @@ struct rom
   }
 };
 
+struct data_retention
+{
+  using result_type = uint32_t;
+  using value_type = uint32_t;
+  
+  static result_type
+  user_get(const xrt_core::device* device)
+  {
+	  return 0;
+  }
+
+  static result_type
+  mgmt_get(const xrt_core::device* device)
+  {
+    auto init_ret = [](const xrt_core::device* dev) {
+    uint32_t ret;
+    mgmtpf::get_data_retention(dev->get_mgmt_handle(), &ret);
+    return ret;
+    };
+
+    static std::map<const xrt_core::device*, uint32_t> info_map;
+    static std::mutex mutex;
+    std::lock_guard<std::mutex> lk(mutex);
+    auto it = info_map.find(device);
+    if (it == info_map.end()) {
+      auto ret = info_map.emplace(device,init_ret(device));
+      it = ret.first;
+    }
+
+    return (*it).second;
+  }
+
+  static void
+  user_put(const xrt_core::device* device, value_type)
+  {
+    throw std::runtime_error("device data retention query is not implemented on user windows");
+  }
+  static void
+    mgmt_put(const xrt_core::device* device, value_type val)
+  {
+    mgmtpf::set_data_retention(device->get_mgmt_handle(), val);
+  }
+
+};
+
+template <typename QueryRequestType, typename Getter>
+struct function0_getput : QueryRequestType
+{
+  static_assert(std::is_same<Getter::result_type, QueryRequestType::result_type>::value
+    || std::is_same<Getter::result_type, boost::any>::value, "get type mismatch");
+  static_assert(std::is_same<Getter::value_type, QueryRequestType::result_type>::value
+    || std::is_same<Getter::value_type, boost::any>::value, "value type mismatch");
+  
+  boost::any
+  get(const xrt_core::device* device) const
+  {
+    if (auto mhdl = device->get_mgmt_handle())
+      return Getter::mgmt_get(device);
+    else if (auto uhdl = device->get_user_handle())
```

Comment:
No need to have "else" statements  (or an "else" tree).   

This is because each "if" statement if true exits the method.